api/v1alpha1/composable_webhook.go

```diff
@@ -0,0 +1,271 @@
+		composablelog.Info("validateApiVersionKind - apiVersion is empty")
+		allErrs = append(allErrs, field.Invalid(fieldpath.Child("apiVersion"), r.Name, "Missing required field - apiVersion"))
+	}
+	if m["kind"] == nil || m["kind"].(string) == "" {
+		composablelog.Info("validateApiVersionKind - kind is empty")
+		allErrs = append(allErrs, field.Invalid(fieldpath.Child("kind"), r.Name, "Missing required field - kind"))
+	}
+	return allErrs
+}
+
+// validate the template content in the spec
+func (r *Composable) validate(template *runtime.RawExtension, fieldpath *field.Path) (map[string]interface{}, field.ErrorList) {
+	var f interface{}
+	json.Unmarshal(template.Raw, &f)
+	m := f.(map[string]interface{})
+	err := r.findGetValueFrom(fieldpath, m)
+	return m, err
+}
+
+// findGetValueFrom parses the template content recursively for getValueFrom elements and validates them
+func (r *Composable) findGetValueFrom(path *field.Path, m map[string]interface{}) field.ErrorList {
+	var allErrs field.ErrorList
+	for k, v := range m {
+		mykey := path.Child(k)
+		switch vv := v.(type) {
+		case string, int32, int64, float32, float64, bool:
+		case []interface{}:
+			if err := r.findGetValueFrom2(mykey, vv); err != nil {
+				allErrs = append(allErrs, err...)
+			}
+		case map[string]interface{}:
+			if vv["getValueFrom"] != nil {
+				newM, _ := json.Marshal(vv["getValueFrom"])
+				myGetValueFrom := ComposableGetValueFrom{}
+				json.Unmarshal(newM, &myGetValueFrom)
+				composablelog.Info("found getValueFrom", "key", "getValueFrom", "value", vv["getValueFrom"])
+				if err := validateGetValueFrom(mykey.String(), myGetValueFrom); err != nil {
+					composablelog.Info("getValueFrom is INVALID", "detail", err)
+					allErrs = append(allErrs, field.Invalid(mykey.Child("getValueFrom"), r.Name, err.Error()))
+				}
+				composablelog.Info("getValueFrom is valid")
+				//set to a random value for dry-run
+				// TODO: set the value to an appropriate type e.g. int, string, etc
+				m[k] = "abc"
+
+			} else { //recursive checking the sub-elements
+				if err := r.findGetValueFrom(mykey, vv); err != nil {
+					allErrs = append(allErrs, err...)
+				}
+			}
+		default:
+			composablelog.Info("findGetValueFrom", "key", mykey, "type-unknown", vv)
+		}
+	}
+	return allErrs
+}
+
+// findGetValueFrom2 functions in the same way as findGetValueFrom above except taking []interface{} as inputs
+// findGetValueFrom2 processes the ararys in the template content for getValueFrom elements
+func (r *Composable) findGetValueFrom2(path *field.Path, m []interface{}) field.ErrorList {
+	var allErrs field.ErrorList
+	for k, v := range m {
+		mykey := path.Child(strconv.Itoa(k))
+		switch vv := v.(type) {
+		case string, int32, int64, float32, float64, bool:
+		case []interface{}:
+			if err := r.findGetValueFrom2(mykey, vv); err != nil {
+				allErrs = append(allErrs, err...)
+			}
+		case map[string]interface{}:
+			if vv["getValueFrom"] != nil {
```

Comment:
The cases processing looks exactly the same as in `findGetValueFrom`, can its implementation be combined in a single function ?

Reply:
The two functions have different input parameter types, one is `map[string]interface{}`, the other is `[]interface{})`.  I may try to  combine them into one that takes both types, will try later.

`findGetValueFrom(path *field.Path, m map[string]interface{})`

`findGetValueFrom2(path *field.Path, m []interface{})`

Reply:
I mean the internal case processing, especially the `case map[string]interface{}`

Reply:
Yes, I can place the processing for `case map[string]interface{}` into a function to reduce the repeating code.

Reply:
consolidated the processing into `func validateGetValueFrom().